### PR TITLE
AMP integration with Holder

### PR DESCRIPTION
**Pull request by AntonOrnatskyi**
Please check the Holder (ukrainian ad network) implementation for amp-ad

I have a Google Individual CLA.
Anton Ornatskyi

### CODE REVIEW

**Comment by AntonOrnatskyi** [comment]
CLA completed

File: ads/_config.js [hunk 1]
```diff
@@ -200,8 +200,13 @@ export const adConfig = {
...
   },
 
-  ibillboard: {},
+  holder: {
+    prefetch: 'https://i.holder.com.ua/js2/holder/ajax/ampv1.js',
+    preconnect: 'https://h.holder.com.ua',
+    renderStartImplemented: true,
+  },
 
+  ibillboard: {},
```
Comment (lannka) [inline ads/_config.js]: nit: keep the blank line after ibillboard

File: ads/holder.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import {writeScript, validateData} from '../3p/3p';
+
+/**
+ * @param {!Window} global
+ * @param {!Object} data
+ */
+export function holder(global, data) {
+  validateData(data, ['block'], ['ampSlotIndex']);
+  const wcl = global.context.location,
+    n = navigator.userAgent;
```
Comment (lannka) [inline ads/holder.js]: either end the previous line with ';', or indent 2 more spaces here.

File: ads/holder.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  let l = '&r' + Math.round((Math.random() * 10000000));
+  l += '&' +
```
Comment (lannka) [inline ads/holder.js]: merge into the previous line?

File: ads/holder.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  'h' + wcl.href;
+  if (!(n.indexOf('Safari') != -1 &&
+  n.indexOf('Chrome') == -1)) {
```
Comment (lannka) [inline ads/holder.js]: move to previous line

File: ads/holder.js (hunk 2 above)
Comment (AntonOrnatskyi) [inline ads/holder.js]: Fixed

File: ads/holder.js (hunk 3 above)
Comment (AntonOrnatskyi) [inline ads/holder.js]: Fixed

File: ads/holder.js (hunk 4 above)
Comment (AntonOrnatskyi) [inline ads/holder.js]: Fixed

**Comment by AntonOrnatskyi** [comment]
/to @lannka Sorry, I put render-start as close to end of ad request as possible now.

File: ads/holder.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+import {writeScript, validateData} from '../3p/3p';
+
+/**
+ * @param {!Window} global
+ * @param {!Object} data
+ */
+export function holder(global, data) {
+  validateData(data, ['block'], ['ampSlotIndex']);
```
Comment (lannka) [inline ads/holder.js]: no need for `ampSlotIndex`. it's a known bug, you can ignore the harmless console error for now.

Comment (AntonOrnatskyi) [inline ads/holder.js]: Done

**Comment by AntonOrnatskyi** [comment]
> I still see a half-second delay between loading indicator and ad render. it is a white blank.

For better performance and for noContentAvailable() realisation I should modify ad-server back-end :( and we'll do it in the one of next updates.

**Comment by AntonOrnatskyi** [comment]
Is it possible to merge PR as is before I'd update back-end?

**Comment by AntonOrnatskyi** [comment]
Thanks I'll add noContentAvailable() and make render-start faster in the week or two.

**Pull request merged**